Mathlib/Topology/Instances/EReal.lean

```diff
@@ -171,6 +170,185 @@ theorem tendsto_nhds_bot_iff_real {α : Type*} {m : α → EReal} {f : Filter α
+    (h : liminf u f = ⊤) (h' : liminf v f ≠ ⊥) : liminf (u + v) f = ⊤ := by
+  apply top_le_iff.1 ((ge_iff_le_forall_real_lt (liminf (u + v) f) ⊤).1 _)
+  intro x
+  rcases exists_between_coe_real (Ne.bot_lt h') with ⟨y, ⟨_, hy⟩⟩
+  intro _
+  rw [← sub_add_cancel x y, coe_add (x - y) y]
+  exact coe_sub x y ▸ @liminf_add_gt_of_gt α f u v (x - y) y
+    (h ▸ coe_sub x y ▸ coe_lt_top (x-y)) hy |>.le
+
+lemma add_liminf_le_liminf_add {α : Type*} {f : Filter α} {u v : α → EReal} :
+    (liminf u f) + (liminf v f) ≤ liminf (u + v) f := by
+  by_cases hu : liminf u f = ⊥
+  · simp_all
+  by_cases hv : liminf v f = ⊥
+  · simp_all
+  have h' : limsup (-(u + v)) f = limsup (-u + -v) f := by
+    apply limsup_congr
+    filter_upwards [eventually_lt_of_lt_liminf (bot_lt_iff_ne_bot.mpr hu),
+      eventually_lt_of_lt_liminf (bot_lt_iff_ne_bot.mpr hv)] with x hux hvx
+    dsimp
+    rw [neg_add (Or.inl hux.ne_bot) (Or.inr hvx.ne_bot), sub_eq_add_neg]
+  rw [← neg_le_neg_iff, neg_add (Or.inl hu) (Or.inr hv), sub_eq_add_neg]
+  rw [← neg_inj, neg_bot] at hu hv
+  simp_rw [← limsup_neg] at hu hv ⊢
+  exact h' ▸ limsup_add_le_add_limsup (Or.inr hv) (Or.inl hu)
+
+lemma limsup_le_iff {α : Type*} {f : Filter α} {u : α → EReal} {b : EReal} :
+    limsup u f ≤ b ↔ ∀ c : ℝ, b < c → ∀ᶠ a : α in f, u a ≤ c := by
+  rw [← le_iff_le_forall_real_gt]
+  refine ⟨?_, ?_⟩ <;> intro h c b_lt_c
+  · rcases exists_between_coe_real b_lt_c with ⟨d, b_lt_d, d_lt_c⟩
+    specialize h d b_lt_d
+    have key := Filter.eventually_lt_of_limsup_lt (lt_of_le_of_lt h d_lt_c)
+    apply Filter.mem_of_superset key
+    rw [Set.setOf_subset_setOf]
+    exact fun a h' ↦ le_of_lt h'
+  · rcases eq_or_neBot f with (rfl | _)
+    · simp only [limsup_bot, bot_le]
+    · specialize h c b_lt_c
+      exact @Filter.limsup_const EReal α _ f _ (c : EReal) ▸ limsup_le_limsup h
+
+lemma limsup_le_const_forall {α : Type*} {f : Filter α} {u : α → EReal} {b : EReal}
+    (h : ∀ a : α, u a ≤ b) : limsup u f ≤ b :=
+  limsup_le_iff.2 fun _ b_lt_c ↦ eventually_of_forall (fun a : α ↦ le_trans (h a) (le_of_lt b_lt_c))
+
+lemma const_le_limsup_forall {α : Type*} {f : Filter α} [NeBot f] {u : α → EReal}
+    {b : EReal} (h : ∀ a : α, b ≤ u a) : b ≤ limsup u f :=
+  @Filter.limsup_const EReal α _ f _ b ▸ limsup_le_limsup (eventually_of_forall h)
+
+lemma liminf_le_const_forall {α : Type*} {f : Filter α} [NeBot f] {u : α → EReal}
+    {b : EReal} (h : ∀ a : α, u a ≤ b) :
+    liminf u f ≤ b :=
+  @Filter.liminf_const EReal α _ f _ b ▸ liminf_le_liminf (eventually_of_forall h)
+
+lemma const_le_liminf_forall {α : Type*} {f : Filter α} {u : α → EReal} {b : EReal}
+    (h : ∀ a : α, b ≤ u a) :
+    b ≤ liminf u f := by
+  rcases eq_or_neBot f with (rfl | _)
+  · simp only [liminf_bot, le_top]
+  · exact @Filter.liminf_const EReal α _ f _ b ▸ liminf_le_liminf (eventually_of_forall h)
+
+lemma limsup_max {α : Type*} {f : Filter α} {u v : α → EReal} :
```

Comment:
Is that specific to EReal, or is it true in a general complete linear order? If so, it would be better to have the general version. (Even better, in a general conditionally complete linear order, with suitable boundedness assumptions that are filled by autoparams, just as is done in the file on limsups and liminfs).

Reply:
@D-Thomine and I are working on it right now. We have decided to remove these lemmas from this PR and proceed as follows in a new PR: 
* Get the more general version of `limsup_le_of_le`
* `limsup_max`: get a version for conditionally complete linear orders (+ autoparam)
* Then write a version for complete linear orders without the autoparam conditions (for the 4 lemmas),
to be put in `Mathlib.Order.LimsupLiminf`